csrc/multidevice/utils.cpp

```diff
@@ -317,30 +317,52 @@ void insertReshardingsAfter(Fusion* fusion) {
   }
 }
 
+void setShardedAllocationDomain(TensorView* tv) {
+  if (!tv->hasAllocation()) {
+    tv->setAllocationDomain(tv->getLeafDomain(), tv->getContiguity());
+  }
+}
+
 } // namespace
 
-void propagateShardings(Fusion* fusion) {
+void propagateShardingsAndSetAllocationDomain(Fusion* fusion) {
+  // Set global input's allocation domain.
+  // Currently we only propagate shardings from consumer to producer,
+  // so a DeviceMesh is required.
+  for (auto global_input_tv :
+       ir_utils::filterByType<TensorView>(fusion->inputs())) {
+    NVF_ERROR(
+        global_input_tv->hasDeviceMesh(),
```

Comment:
This is just a requirement that we choose to enforce, right? so probably needs to be NVF_CHECK instead.